share/source.F90

```diff
@@ -12,35 +12,58 @@
 ! Full instructions on compiling and using MCNP5 with this subroutine are found
 ! in the PyNE user manual.
 
-function find_cell() result(icl_tmp)
-! This function to determines the current MCNP cell index location and exits if
-! no valid cell is found. This only works if there are no repeated geometries or
-! universes present in the model.
+function find_cell(cell_list, cell_list_size) result(icl_tmp)
+! This function to determines the current MCNP cell index location.
+! Return a positive integer if a valid cell is found, otherwise it returns -1.
+! This only works if there are no repeated geometries or universes present in
+! the model.
+! Parameters:
+!     cell_list: array of integers. Contains cell numbers that the source
+!                particle possiblely located.
+!     cell_list_size: integer. Size of the cell_list.
 
     use mcnp_global
     use mcnp_debug
     ! xxx,yyy,zzz are global variables
     ! mxa is global
     integer :: i ! iterator variable
-    integer :: j ! tempory cell test
+    integer :: j ! temporary cell test
     integer :: icl_tmp ! temporary cell variable
-    icl_tmp = -1
-
-    do i = 1, mxa
-      call chkcel(i, 0, j)
-      if (j .eq. 0) then
-         ! valid cel set
-         icl_tmp = i
-         exit
-      endif
-    enddo
-    ! icl is now set
+    integer, intent(in) :: cell_list_size
+    integer, dimension(cell_list_size), intent(in) :: cell_list
+    integer :: cid ! cell index
 
-    if(icl_tmp .le. 0) then
-      write(*,*) 'Nonsense cell number stopping'
-      stop
+    icl_tmp = -1
+    if (cell_list_size .eq. 0) then
+        ! TET mesh
+        do i = 1, mxa
+           call chkcel(i, 0, j)
+           if (j .eq. 0) then
+              ! valid cell found
+              icl_tmp = i
+              exit
+           endif
+        enddo
+    else
+        ! HEX mesh. VOXEL/SUBVOXEL
+        do i = 1, cell_list_size
+           if (cell_list(i) .le. 0) then
+               ! not a valid cell number (-1)
+               exit
+           endif
+           cid = namchg(1, cell_list(i))
+           if (cid .eq. 0) then
```

Comment:
Should this case ever be hit? Should we be halting the program here?

Reply:
This line is hit by a tiny chance in FNG model, it is firstly hit at about nps between 1e6~1e7.
This line is not hit if I use a very simple geometry (test case, rectangular cells, less than 10 cells).

I didn't halt the program at this line before, but I think it worth to do it.

Reply:
My sense is that this should never happen. And if we do hit that case, there is something systematic about what's going wrong. We should always be able to look up the index of a valid cell in the problem right?

Reply:
Yeah, I think I can repeat that problem with more tests, and then find out the reason.

Reply:
I repeated this problem and found that when the code hit this line, the cell number is 138. I checked the output file of both neutron transport and photon transport, found that this cell 138 exist in neutron transport but not exist in photon transport.
The material of this cell 138 is air.
It reminds me that the model for neutron transport and photon transport are slightly different:
- The lateral port plug exists in the model for neutron transport, but not in the model for photon transport
- Two detectors exist in the model for photon transport, but not neutron transport
I will double check the location of this cell 138 tomorrow.

Reply:
The vol 138 is a thin layer of air around detectors. It is removed in photon transport, it won't influence the results.